external/storm-kafka-client/src/test/java/org/apache/storm/kafka/bolt/KafkaBoltTest.java

```diff
@@ -17,73 +17,140 @@
-            }
-        }), any(Callback.class));
+
+        assertThat(producer.history().size(), is(1));
+        ProducerRecord<String, String> arg = producer.history().get(0);
+
+        LOG.info("GOT {} ->", arg);
+        LOG.info("{}, {}, {}", arg.topic(), arg.key(), arg.value());
+        assertThat(arg.topic(), is("MY_TOPIC"));
+        assertThat(arg.key(), is(key));
+        assertThat(arg.value(), is(value));
+
+        // Complete the send
+        producer.completeNext();
         verify(collector).ack(testTuple);
     }
 
+    @Test
+    public void testSimpleWithError() {
+        MockProducer<String, String> producer = new MockProducer<>(Cluster.empty(), false, null, null, null);
+        KafkaBolt<String, String> bolt = makeBolt(producer);
+
+        OutputCollector collector = mock(OutputCollector.class);
+        TopologyContext context = mock(TopologyContext.class);
+        Map<String, Object> conf = new HashMap<>();
+        bolt.prepare(conf, context, collector);
+
+        String key = "KEY";
+        String value = "VALUE";
+        Tuple testTuple = createTestTuple(key, value);
+        bolt.execute(testTuple);
+
+        assertThat(producer.history().size(), is(1));
+        ProducerRecord<String, String> arg = producer.history().get(0);
+
+        LOG.info("GOT {} ->", arg);
+        LOG.info("{}, {}, {}", arg.topic(), arg.key(), arg.value());
+        assertThat(arg.topic(), is("MY_TOPIC"));
+        assertThat(arg.key(), is(key));
+        assertThat(arg.value(), is(value));
+
+        // Force a send error
+        KafkaException ex = new KafkaException();
+        producer.errorNext(ex);
+        verify(collector).reportError(ex);
+        verify(collector).fail(testTuple);
+    }
+
+    @Test
+    public void testCustomCallbackIsWrappedByDefaultCallbackBehavior() {
+        MockProducer<String, String> producer = new MockProducer<>(Cluster.empty(), false, null, null, null);
+        KafkaBolt<String, String> bolt = makeBolt(producer);
+
+        PreparableCallback customCallback = mock(PreparableCallback.class);
+        bolt.withProducerCallback(customCallback);
+
+        OutputCollector collector = mock(OutputCollector.class);
+        bolt.prepare(new HashMap<>(), mock(TopologyContext.class), collector);
+
+        String key = "KEY";
+        String value = "VALUE";
+        Tuple testTuple = createTestTuple(key, value);
+        bolt.execute(testTuple);
+
+        assertThat(producer.history().size(), is(1));
+        ProducerRecord<String, String> arg = producer.history().get(0);
+
+        LOG.info("GOT {} ->", arg);
+        LOG.info("{}, {}, {}", arg.topic(), arg.key(), arg.value());
+        assertThat(arg.topic(), is("MY_TOPIC"));
+        assertThat(arg.key(), is(key));
+        assertThat(arg.value(), is(value));
+
+        // Complete the send
+        producer.completeNext();
+        verify(collector).ack(testTuple);
+        verify(customCallback).onCompletion(any(), any());
```

Comment:
Nit: Can we be more specific about the parameters here, or would that be hard to do?

Reply:
Do you mean in regards to the `any()`? I'll see if I can make it a little better.

Reply:
Yes. It's nitpicking though, so if it's not super easy don't worry about it. Either way, please squash to one commit when you want this merged.

Reply:
No problem.

Reply:
I was able to make it a little better. I went ahead and squashed the commits and it should be okay to merge now.